google/cloud/bigtable/_read_rows.py

```diff
@@ -0,0 +1,683 @@
+        """
+        if self._last_emitted_row_key is not None:
+            # if this is a retry, try to trim down the request to avoid ones we've already processed
+            try:
+                self._request["rows"] = _ReadRowsOperation._revise_request_rowset(
+                    row_set=self._request.get("rows", None),
+                    last_seen_row_key=self._last_emitted_row_key,
+                )
+            except _RowSetComplete:
+                # if there are no rows left to process, we're done
+                return
+            # revise next request's row limit based on number emitted
+            if total_row_limit:
+                new_limit = total_row_limit - self._emit_count
+                if new_limit <= 0:
+                    return
+                else:
+                    self._request["rows_limit"] = new_limit
+        params_str = f'table_name={self._request.get("table_name", "")}'
+        if self._request.get("app_profile_id", None):
+            params_str = (
+                f'{params_str},app_profile_id={self._request.get("app_profile_id", "")}'
+            )
+        time_to_deadline = operation_deadline - time.monotonic()
+        gapic_timeout = min(time_to_deadline, per_request_timeout)
+        new_gapic_stream = await gapic_fn(
+            self._request,
+            timeout=gapic_timeout,
+            metadata=[("x-goog-request-params", params_str)],
+        )
+        buffer: asyncio.Queue[Row | RequestStats | Exception] = asyncio.Queue(
+            maxsize=buffer_size
+        )
+        buffer_task = asyncio.create_task(
+            self._generator_to_buffer(buffer, new_gapic_stream)
+        )
+        buffered_stream = self._buffer_to_generator(buffer)
+        state_machine = _StateMachine()
+        try:
+            stream = _ReadRowsOperation.merge_row_response_stream(
+                buffered_stream, state_machine
+            )
+            # run until we get a timeout or the stream is exhausted
+            while True:
+                new_item = await stream.__anext__()
+                # ignore rows that have already been emitted
+                if isinstance(new_item, Row):
+                    if self._last_emitted_row_key is not None and new_item.row_key <= self._last_emitted_row_key:
+                        raise InvalidChunk("Last emitted row key out of order")
+                    # don't yeild _LastScannedRow markers; they
+                    # should only update last_seen_row_key
+                    if not isinstance(new_item, _LastScannedRow):
+                        yield new_item
+                        self._emit_count += 1
+                    self._last_emitted_row_key = new_item.row_key
+                    if total_row_limit and self._emit_count >= total_row_limit:
+                        return
+                elif isinstance(new_item, RequestStats):
+                    yield new_item
```

Comment:
I think we should have special handling for RequestStats:
- RequestStats are almost always emited once at the end, except during race conditions with multiple retry attempts
- Mixing metadata and data in a single stream feels like we are mixing concerns
- Having RequestStats in a type signature when it's only present when a request bit is set seems a bit sloppy

What do you think of: passing a RequestStatsCollector as an output method parameter to ReadRows. When its present, it will set the appropriate proto field and populate at the end of the operation

Reply:
Yeah that idea sounds like it could be a good option to me, I'll look into it. My initial thought was to expose request_stats as an attribute of the generator we pass to the user, since we already needed a custom generator object for other reasons. But that doesn't work as well for the sync surface, since I don't want to also make a custom list subclass. So a RequestStatsCollector might be better for that reason

Is it alright if we address this in a separate PR though? My goal here was to just set up some of the plumbing for the request_stats here, and add the main user-facing implementation later. Maybe I should just remove the references to request_stats from here for now?

Reply:
yes, please remove it. My memory for these things is terrible :(